test/e2e/runner/load.go

```diff
@@ -27,73 +25,101 @@ func Load(ctx context.Context, testnet *e2e.Testnet, multiplier int) error {
 	if concurrency == 0 {
 		concurrency = 1
 	}
-	initialTimeout := 1 * time.Minute
-	stallTimeout := 30 * time.Second
 
 	chTx := make(chan types.Tx)
-	chSuccess := make(chan types.Tx)
+	chSuccess := make(chan int)
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 
 	// Spawn job generator and processors.
 	logger.Info(fmt.Sprintf("Starting transaction load (%v workers)...", concurrency))
 	started := time.Now()
 
-	go loadGenerate(ctx, chTx, multiplier, testnet.TxSize)
+	go loadGenerate(ctx, chTx, testnet.TxSize)
 
 	for w := 0; w < concurrency; w++ {
 		go loadProcess(ctx, testnet, chTx, chSuccess)
 	}
 
-	// Monitor successful transactions, and abort on stalls.
+	// Montior transaction to ensure load propagates to the network
+	//
+	// note: previous versions of this loop ensured that the load
+	// didn't stall for more than 30 seconds, but checking for
+	// stalls here just aborts the load generator earlier, and can
+	// easily reflect backpressure in the test harness. What we
+	// should care about more than stalls is, 80th or 95th
+	// percentile latency, and this is not feasible to track here
+	// right now.
 	success := 0
-	timeout := initialTimeout
 	for {
 		select {
-		case <-chSuccess:
-			success++
-			timeout = stallTimeout
-		case <-time.After(timeout):
-			return fmt.Errorf("unable to submit transactions for %v", timeout)
+		case numSeen := <-chSuccess:
+			success += numSeen
 		case <-ctx.Done():
 			if success == 0 {
 				return errors.New("failed to submit any transactions")
 			}
-			logger.Info(fmt.Sprintf("Ending transaction load after %v txs (%.1f tx/s)...",
-				success, float64(success)/time.Since(started).Seconds()))
+			rate := float64(success) / time.Since(started).Seconds()
+
+			logger.Info("ending transaction load",
+				"dur_secs", time.Since(started).Seconds(),
+				"txns", success,
+				"rate", rate)
+
+			if rate < 2.0 {
+				logger.Error("transaction throughput was low",
+					"rate", rate)
+			}
+
 			return nil
 		}
 	}
 }
 
-// loadGenerate generates jobs until the context is canceled
-func loadGenerate(ctx context.Context, chTx chan<- types.Tx, multiplier int, size int64) {
-	for i := 0; i < math.MaxInt64; i++ {
+// loadGenerate generates jobs until the context is canceled.
+//
+// The chTx has multiple consumers, thus the rate limiting of the load
+// generation is primarily the result of backpressure from the
+// broadcast transaction, though at most one transaction will be
+// produced every 10ms.
+func loadGenerate(ctx context.Context, chTx chan<- types.Tx, size int64) {
+	timer := time.NewTimer(0)
```

Comment:
Was this meant to be 10ms as described?
```suggestion
	timer := time.NewTimer(10*time.Millisecond)
```

Edit: Oh, no, I see what's going on here. You want the first one to fire immediately.

Perhaps:

```go
// This timer is reset below after a tx is received. We want the initial
// tick to happen immediately, however.
```

Reply:
I feel like `time.NewTimer(0)` is pretty clear for "I want this to tick immediately, and will reset it to what I need later..." otherwise you'd just use a ticker or set something differently?

Reply:
Having seen the usage, yes—but most of the time when I see something like this I assume it's a typo. 🙂